### PR TITLE
Add BOM substitution and tests for Sass and SCSS filters

### PR DESCRIPTION
**Pull request by benburwell**
Addresses #3914. The tests feel a little clunky, what with creating a new JekyllFilter, but I didn't want to convert the existing test cases to `:compressed` mode without an explicit go-ahead/request to do so.

### CODE REVIEW

**Comment by envygeeks** [comment]
I don't know that we should add `@charset` more than just entirely strip it, unless there is something I'm missing?   Other than that :shipit:

**Comment by parkr** [comment]
Fixes #3914.

**Comment by parkr** [comment]
> I don't know that we should add `@charset`

What's the rationale here?

**Comment by benburwell** [comment]
Rationale for adding the charset was that the BOM is intentionally inserted by Sass instead of the charset rule under particular circumstances to signal to the CSS client that it's a UTF-8 encoding. With the substitution, the charset rule would only be added if a BOM is removed.

If we are guaranteed to be generating a UTF-8 file, (with a BOM itself, or a HTML meta tag, or served with that encoding header), we could safely drop this according to (my reading of) the spec.

**Comment by parkr** [comment]
Is this a setting of the Sass converter? Maybe we could tweak it over at jekyll-sass-converter.

I could see people not wanting the charset injected into their snippets

**Comment by benburwell** [comment]
Basically, when the Sass/SCSS source contains non-ASCII characters, [Sass inserts either a BOM or the charset rule](https://github.com/sass/sass/blob/master/lib/sass/tree/visitors/to_css.rb#L131). If the compressed mode is being used, it's the BOM, otherwise it uses the charset rule.

So the only way there would be a BOM to begin with is if the file contains non-ASCII characters, in which case Sass would be adding the charset rule anyway.

**Comment by parkr** [comment]
I'm fine with adding in the charset if we think they'll only end up in CSS files or style blocks. I'd hate to have someone have some highlight block that has that header thrown in when they don't want it.

**Comment by benburwell** [comment]
That makes sense. I'm not too familiar with all of Jekyll's internals -- is there a place to differentiate in what context the converter is being called? I looked around a bit but I didn't find anything. May easily have missed it though.

**Comment by parkr** [comment]
> is there a place to differentiate in what context the converter is being called?

No sir, this is not available. So we'll have to choose something and decide it fits for all circumstances. I don't see `scssify` being used in snippets all that often, so let's just go ahead with what you have and we'll see what feedback you get.

@envygeeks, any thoughts?

**Comment by envygeeks** [comment]
Jekyll filters support multiple arguments so we could always create an undocumented API argument to switch the context and if people notice it... they can abuse it otherwise it's there for us to differentiate internally without having to dup code.

**Comment by envygeeks** [comment]
Any update on this? We are coming to close of the 3.0 window so if you want to see this in the release.

**Comment by parkr** [comment]
> Jekyll filters support multiple arguments so we could always create an undocumented API argument to switch the context and if people notice it... they can abuse it otherwise it's there for us to differentiate internally without having to dup code.

They're instantiated with the site config, so that could be used here. Nevertheless, let's strip the BOM. Add an option to include the CHARSET declaration.

**Comment by benburwell** [comment]
Okay, I'll update the PR to use a config option for whether to add the charset rule. Probably not until tomorrow night, but possibly later this evening.

**Comment by envygeeks** [comment]
Just noticed it was updated /cc @jekyll/core -- Should be ready for re-review.  I'm :shipit:

File: lib/jekyll/configuration.rb [hunk 1]
```diff
@@ -38,6 +38,7 @@ class Configuration < Hash
       'highlighter'   => 'rouge',
       'lsi'           => false,
       'excerpt_separator' => "\n\n",
+      'add_charset'   => true,
```
Comment (parkr) [inline lib/jekyll/configuration.rb]: What do you think about making this a sass rule?

```yaml
sass:
  style: compressed
  add_charset: true
```

then we can take care of it over in jekyll-sass-converter. i think this shouldnt' be in the core now that i have had more time to think on it.

**Comment by benburwell** [comment]
I agree. I was actually thinking about that last night when I was putting this in. I'll submit a new PR in jekyll-sass-converter tonight. I think this one can be closed then.

**Comment by parkr** [comment]
Thanks, @benburwell!

**Pull request closed without merging**